File: Lowest Common Ancestor in Binary Tree/Leetcode_236.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        if root in (p, q):
            return root

        left = self.lowestCommonAncestor(root.left, p, q) if root.left else None
        right = self.lowestCommonAncestor(root.right, p, q) if root.right else None

        return root if left and right else left if left else right

# same algo, return namedtuple instead of directly

from collections import namedtuple

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        Status = namedtuple('Status','numTargets ancestor')
        
        def lca_helper(tree, p, q):
            if tree is None: return Status(0, None)
            l = lca_helper(tree.left, p, q)
            if l.numTargets==2: return l
            r = lca_helper(tree.right, p, q)
            if r.numTargets==2: return r
            numTargets = (l.numTargets + r.numTargets + (p,q).count(tree))
            return Status(numTargets, tree if numTargets==2 else None)
        
        return lca_helper(root, p, q).ancestor

# brute force solution
# O(n^2) runtime

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root in (None, p, q): return root
        left, right = (self.lowestCommonAncestor(kid, p, q) for kid in (root.left, root.right))
        return root if left and right else left or right

File: Lowest Common Ancestor in Binary Tree/Leetcode_236.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// O(N) time, O(h) space
// a-la postorder traversal
// slower on LC than recursive (presumably b/c overhead)

class Solution {
private:
    typedef struct Status {
        Status(): numTargets(0), ancestor(NULL) {}
        int numTargets;
        TreeNode* ancestor;
    } Status;
    
    Status* lca_helper(TreeNode* root, TreeNode* p, TreeNode* q){
        Status* curStatus = new Status;
        if (root == NULL) return curStatus; 
            
        Status* left = lca_helper(root->left, p, q);
        if (left->numTargets == 2) return left;
        else curStatus->numTargets += left->numTargets;
        
        Status* right = lca_helper(root->right, p, q);
        if (right->numTargets == 2) return right;
        else curStatus->numTargets += right->numTargets;
     
        if ((p->val == root->val) || (q->val == root->val)) curStatus->numTargets += 1;
        if (curStatus->numTargets == 2) curStatus->ancestor = root;
        return curStatus;
    };

public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return lca_helper(root, p, q)->ancestor;
    }
};

// O(N^2) time, O(h) space
// recursion
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if ((root == NULL) || (root==p) || (root==q)) return root;
        TreeNode* left = lowestCommonAncestor(root->left, p, q);
        TreeNode* right = lowestCommonAncestor(root->right, p, q);
        if ((left != NULL) && (right != NULL)){
            return root;
        } else if (left != NULL) return left;
        else return right;
    }
};

